File: v2/src/main/java/com/api/v2/customers/controllers/CustomerController.java
package com.api.v2.customers.controllers;

import com.api.v2.customers.responses.CustomerResponseDto;
import com.api.v2.customers.services.CustomerRegistrationService;
import com.api.v2.customers.services.CustomerUpdatingService;
import com.api.v2.people.requests.PersonRegistrationDto;
import com.api.v2.people.requests.PersonUpdatingDto;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("api/v2/customers")
@RequiredArgsConstructor
public class CustomerController {

    private final CustomerRegistrationService registrationService;
    private final CustomerUpdatingService updatingService;

    @PostMapping
    public Mono<ResponseEntity<CustomerResponseDto>> register(@Valid @RequestBody PersonRegistrationDto registrationDto) {
        return registrationService.register(registrationDto);
    }

    @PatchMapping("{customerId}")
    public Mono<ResponseEntity<Void>> update(@PathVariable String customerId, @Valid @RequestBody PersonUpdatingDto updatingDto) {
        return updatingService.update(customerId, updatingDto);
    }
}

File: v2/src/main/kotlin/com/api/v2/customers/controllers/CustomerController.kt
package com.api.v2.customers.controllers

import com.api.v2.customers.CustomerResponseDTO
import com.api.v2.customers.services.*
import com.api.v2.people.PersonRegistrationDTO
import com.api.v2.people.PersonUpdateDTO
import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("api/v2/customers")
class CustomerController(
    private val registrationService: CustomerRegistrationService,
    private val updateService: CustomerUpdateService,
    private val retrievalService: CustomerRetrievalService
) {

    @PostMapping
    fun register(@RequestBody registrationDTO: @Valid PersonRegistrationDTO): ResponseEntity<CustomerResponseDTO> {
        return registrationService.register(registrationDTO)
    }

    @PatchMapping("{customerId}")
    fun update(
        @PathVariable customerId: String,
        @RequestBody updateDTO: @Valid PersonUpdateDTO
    ): ResponseEntity<Unit> {
        return updateService.update(customerId, updateDTO)
    }

    @GetMapping("{id}")
    fun findById(@PathVariable id: String): ResponseEntity<CustomerResponseDTO> {
        return retrievalService.findById(id)
    }

    @GetMapping
    fun findAll(pageable: Pageable): ResponseEntity<Page<CustomerResponseDTO>> {
        return retrievalService.findAll(pageable)
    }

}